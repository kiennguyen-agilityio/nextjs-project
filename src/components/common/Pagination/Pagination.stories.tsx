import type { Meta, StoryObj } from '@storybook/react';

// components
import Pagination from '@/components/common/Pagination';

const meta = {
  title: 'Components/Pagination',
  component: Pagination,
  parameters: {
    layout: 'centered',
  },
} satisfies Meta<typeof Pagination>;

export default meta;

type Story = StoryObj<typeof Pagination>;

export const Default: Story = {
  args: {
    totalUsers: 10,
    limit: 5,
    currentPage: 1,
  },
};
