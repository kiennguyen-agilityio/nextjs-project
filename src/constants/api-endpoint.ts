export const API_ENDPOINT = {
  CATEGORY_LIST: 'categories',
  USER_LIST: 'users',
  ROLE_LIST: 'roles',
};
