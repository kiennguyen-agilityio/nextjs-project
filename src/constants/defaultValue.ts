export const DEFAULT_LIMIT = 4;
export const DEFAULT_DURATION = 3000;
export const DEFAULT_AVATAR_URL =
  'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcRyKLQ_NDd81udMvX8pB7D97hkZxbjehU6WzA&s';
export const DEFAULT_ROW_CLASS = 'py-2 px-4 border-b border-r border-gray-200';
