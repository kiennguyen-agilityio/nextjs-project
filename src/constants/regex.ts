export const REGEX = {
  IMG: /\.(png|jpg|webp)$/,
};
