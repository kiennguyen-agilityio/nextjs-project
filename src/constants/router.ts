export const ROUTER = {
  CATEGORY: '/category',
  LOGIN: '/login',
  ROLES: '/roles',
  USERS: '/users',
};

// Use ROUTES in LINKS
export const LINKS = [
  {
    name: 'Roles',
    href: ROUTER.ROLES,
  },
  {
    name: 'Users',
    href: ROUTER.USERS,
  },
];
