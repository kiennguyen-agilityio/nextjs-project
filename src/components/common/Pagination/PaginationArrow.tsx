import Link from 'next/link';

// icons
import { ChevronRightIcon, ChevronLeftIcon } from '@/icons';

// components
import { Button } from '@/components/common/Button';

interface PaginationArrowProps {
  href: string;
  direction: 'left' | 'right';
  isDisabled?: boolean;
}

const PaginationArrow = ({
  href,
  direction,
  isDisabled,
}: PaginationArrowProps) => {
  const baseClass =
    'flex h-8 w-8 items-center justify-center rounded border-none';
  const disabledClass = isDisabled
    ? 'pointer-events-none text-gray-300'
    : 'hover:bg-gray-200';
  const marginClass = direction === 'left' ? 'mr-2' : 'ml-2';

  const className = `${baseClass} ${disabledClass} ${marginClass}`;

  const icon =
    direction === 'left' ? <ChevronLeftIcon /> : <ChevronRightIcon />;

  return isDisabled ? (
    <Button customClass={className} aria-disabled="true" variant="outline">
      {icon}
    </Button>
  ) : (
    <Link href={href} className={className}>
      {icon}
    </Link>
  );
};

export default PaginationArrow;
