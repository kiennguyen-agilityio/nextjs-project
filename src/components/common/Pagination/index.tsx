'use client';
import { usePathname, useSearchParams } from 'next/navigation';

// components
import PaginationArrow from '@/components/common/Pagination/PaginationArrow';

// constants
import { DEFAULT_LIMIT } from '@/constants/defaultValue';

interface PaginationProps {
  totalUsers: number;
  limit: number;
  currentPage: number;
}

export default function Pagination({
  totalUsers,
  limit,
  currentPage,
}: PaginationProps) {
  const pathname = usePathname();
  const searchParams = useSearchParams();
  const startItem = (currentPage - 1) * limit + 1;
  const endItem = Math.min(currentPage * limit, totalUsers);
  const totalPages = Math.ceil(totalUsers / limit);

  const createPageURL = (pageNumber: number | string) => {
    const params = new URLSearchParams(searchParams);
    params.set('page', pageNumber.toString());
    params.set('limit', limit.toString());
    return `${pathname}?${params.toString()}`;
  };

  const itemsPerPage = DEFAULT_LIMIT;

  return (
    <div className="flex items-center justify-between">
      <div className="text-sm">
        Displaying <strong>{itemsPerPage}</strong> out of {totalPages}
      </div>

      <div className="flex items-center">
        <div className="flex items-center mx-2">
          <span className="text-sm font-medium">{startItem}</span>
          <span className="mx-1">-</span>
          <span className="text-sm font-medium">{endItem}</span>
        </div>
        <PaginationArrow
          direction="left"
          href={createPageURL(currentPage - 1)}
          isDisabled={currentPage <= 1}
        />

        <PaginationArrow
          direction="right"
          href={createPageURL(currentPage + 1)}
          isDisabled={currentPage >= totalPages}
        />
      </div>
    </div>
  );
}
