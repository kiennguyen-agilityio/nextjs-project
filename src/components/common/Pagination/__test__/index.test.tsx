import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// constants
import { DEFAULT_LIMIT } from '@/constants/defaultValue';

// components
import Pagination from '@/components/common/Pagination';

const mockUsePathname = jest.fn();
const mockUseSearchParams = jest.fn();

jest.mock('next/navigation', () => ({
  usePathname: () => mockUsePathname(),
  useSearchParams: () => mockUseSearchParams(),
}));

describe('Pagination Component', () => {
  beforeEach(() => {
    mockUsePathname.mockReturnValue('/test');
    mockUseSearchParams.mockReturnValue(new URLSearchParams());
  });

  it('matches snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <Pagination totalUsers={100} limit={DEFAULT_LIMIT} currentPage={1} />
      </BrowserRouter>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
