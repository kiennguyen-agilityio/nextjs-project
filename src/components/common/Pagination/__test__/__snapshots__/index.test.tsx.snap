// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center justify-between"
  >
    <div
      class="text-sm"
    >
      Displaying 
      <strong>
        2
      </strong>
       out of 50
    </div>
    <div
      class="flex items-center"
    >
      <div
        class="flex items-center mx-2"
      >
        <span
          class="text-sm font-medium"
        >
          1
        </span>
        <span
          class="mx-1"
        >
          -
        </span>
        <span
          class="text-sm font-medium"
        >
          2
        </span>
      </div>
      <button
        class="flex items-center rounded p-2 border border-gray-200 dark:border-gray-700 hover:bg-gray-100 hover:opacity-80 flex h-8 w-8 items-center justify-center rounded border-none pointer-events-none text-gray-300 mr-2"
        type="button"
      >
        <svg
          class="w-4 h-4"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.75 19.5 8.25 12l7.5-7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
      <a
        class="flex h-8 w-8 items-center justify-center rounded border-none hover:bg-gray-200 ml-2"
        href="/test?page=2&limit=2"
      >
        <svg
          class="w-4 h-4"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8.25 4.5 7.5 7.5-7.5 7.5"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
